src/opuntia.erl

```diff
@@ -132,27 +146,49 @@ calculate(#token_bucket_shaper{shape = {MaximumTokens, Rate},
     TokensOverused = max(0, TokensNowUsed - ExactlyAvailableNow),
 
     %% And then MaybeDelay will be zero if TokensOverused was zero
-    MaybeDelayMs = TokensOverused / Rate,
+    OverUsedRateNow = TokensOverused / Rate,
 
     %% We penalise rounding up, the most important contract is that rate will never exceed that
     %% requested, but the same way timeouts in Erlang promise not to arrive any time earlier but
     %% don't promise at what time in the future they would arrive, nor we promise any upper bound
     %% to the limits of the shaper delay.
-    RoundedDelayMs = ceil(MaybeDelayMs),
-
-    NewShaper = Shaper#token_bucket_shaper{available_tokens = TokensAvailable,
-                                           last_update = NativeNow + RoundedDelayMs,
-                                           debt = RoundedDelayMs - MaybeDelayMs},
-    {NewShaper, RoundedDelayMs}.
+    case OverUsedRateNow - OverPenalisedInUnitsLastTime of
```

Comment:
IMO this should be a separate function with two clauses. Some of the comments could be replaced with meaningful function names. This single clause of `calculate` is 50 lines now (!).